components/feature/awesomebar/src/main/java/mozilla/components/feature/awesomebar/provider/ClipboardSuggestionProvider.kt

```diff
@@ -0,0 +1,60 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.awesomebar.provider
+
+import android.content.ClipboardManager
+import android.content.Context
+import mozilla.components.concept.awesomebar.AwesomeBar
+import mozilla.components.feature.session.SessionUseCases
+import mozilla.components.support.utils.WebURLFinder
+
+private const val MIME_TYPE_TEXT_PLAIN = "text/plain"
+
+/**
+ * An [AwesomeBar.SuggestionProvider] implementation that returns a suggestions for an URL in the clipboard (if there's
+ * any).
+ */
+class ClipboardSuggestionProvider(
+    context: Context,
+    private val loadUrlUseCase: SessionUseCases.LoadUrlUseCase
+) : AwesomeBar.SuggestionProvider {
+    private val clipboardManager = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
+
+    override suspend fun onInputChanged(text: String): List<AwesomeBar.Suggestion> {
+        val url = getTextFromClipboard(clipboardManager)?.let {
+            findUrl(it)
+        } ?: return emptyList()
+
+        return listOf(AwesomeBar.Suggestion(
+            id = "mozac-feature-awesomebar-clipboard",
+            description = url,
+            flags = setOf(AwesomeBar.Suggestion.Flag.CLIPBOARD),
+            onSuggestionClicked = {
+                loadUrlUseCase.invoke(url)
+            }
+        ))
+    }
+}
+
+private fun findUrl(text: String): String? {
+    val finder = WebURLFinder(text)
+    return finder.bestWebURL()
+}
+
+@Suppress("ReturnCount")
+private fun getTextFromClipboard(clipboardManager: ClipboardManager): String? {
+    if (!clipboardManager.hasPrimaryClip()) {
+        return null
+    }
+
+    if (clipboardManager.primaryClipDescription?.hasMimeType(MIME_TYPE_TEXT_PLAIN) != true) {
+        return null
+    }
+
+    val item = clipboardManager.primaryClip?.getItemAt(0)
+        ?: return null
+
+    return item.text?.toString()
```

Comment:
I found this function a bit difficult to read, because of multiple returns. Maybe smaller functions could help :)

```kotlin
private fun getTextFromClipboard(clipboardManager: ClipboardManager): String? {
    return with(clipboardManager) {
        if (hasNotPlainText() or isPrimaryEmpty()) {
            null
        } else {
            first().toString()
        }
    }
}

fun ClipboardManager.hasNotPlainText(): Boolean {
    return !(primaryClipDescription?.hasMimeType(MIME_TYPE_TEXT_PLAIN) ?: false)
}

fun ClipboardManager.isPrimaryEmpty(): Boolean {
    return primaryClip?.itemCount == 0
}

fun ClipboardManager.first(): ClipData.Item? {
    return primaryClip?.getItemAt(0)
}
```